### PR TITLE
Migrate recent community changes

### LINKED ISSUE #17340: Review community-project latest changes
Review latest changes on the community project (https://github.com/camunda-community-hub/spring-zeebe/pull/690) and move them into the client if necessary

### PR DESCRIPTION
**Pull request by ana-vinogradova-camunda**
## Description

Migrating recents community project changes.
More details in [this issue](https://github.com/camunda/zeebe/issues/17340)

The main change is in how application properties are handled. Properties are read from .yaml files (application-camunda-<mode>.yaml) instead of using multiple configuration classes. As a result, multiple configuration files were marked as deprecated or removed and classes that are used for properties handling were added.

## Related issues
closes #17340

### CODE REVIEW

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/CommonClientConfiguration.java [hunk 1]
```diff
@@ -44,6 +46,7 @@ public class CommonClientConfiguration {
   ZeebeSelfManagedProperties zeebeSelfManagedProperties;
 
   @Bean
+  @ConditionalOnMissingBean
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/CommonClientConfiguration.java:49]: ❓ Why this annotation is needed ?

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+   *
+   * @param propertyName the name of the property, used for logging
+   * @param propertySupplier a function to supply the property, may throw
+   * @param legacyPropertySupplier a function to supply the legacy property, may throw
+   * @param defaultProperty the default to apply if nothing else suits, may be null
+   * @param configCache the cache to save the property to, may be null
+   * @return the property resolved
+   * @param <T> the type of the property
+   */
+  public static <T> T getOrLegacyOrDefault(
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java:38]: 🔧 I suggest to refactor a bit this method, is a bit confusing

```suggestion
  public static <T> T getOrLegacyOrDefault(
    final String propertyName,
    final Supplier<T> propertySupplier,
    final Supplier<T> legacyPropertySupplier,
    final T defaultProperty,
    final Map<String, Object> configCache) {

    if (configCache != null && configCache.containsKey(propertyName)) {
        LOG.debug("Property {}: Loading from cache", propertyName);
        return (T) configCache.get(propertyName);
    }

    T property = defaultProperty;
    if (property == null) {
        property = getPropertyFromSupplier(legacyPropertySupplier, propertyName, "legacy");
    }
    if (property == null || property.equals(defaultProperty)) {
        property = getPropertyFromSupplier(propertySupplier, propertyName, "property");
    }
    if (property == null || property.equals(defaultProperty)) {
        LOG.debug("Property {}: not set or default, using default", propertyName);
        property = defaultProperty;
    }
    if (configCache != null) {
        configCache.put(propertyName, property);
    }
    return property;
}

private static <T> T getPropertyFromSupplier(
    Supplier<T> supplier,
    String propertyName,
    String propertyType) {

    T property = null;
    try {
        LOG.debug("Property {}: not set or default, applying {} property", propertyName, propertyType);
        property = supplier.get();
    } catch (final Exception e) {
        LOG.debug("Error while loading {} property {}", propertyType, propertyName, e);
    }
    return property;
}
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java [hunk 3]
```diff
@@ -175,43 +277,86 @@ public ScheduledExecutorService jobWorkerExecutor() {
...
 
   @Override
   public boolean preferRestOverGrpc() {
-    return false;
+    return camundaClientProperties.getZeebe().isPreferRestOverGrpc();
   }
 
-  private CredentialsProvider initCredentialsProvider() {
+  private CredentialsProvider identityCredentialsProvider() {
+    if (!(authentication instanceof DefaultNoopAuthentication)) {
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:309]: 🔧 
```suggestion
    if (authentication instanceof DefaultNoopAuthentication) {
        return null;
    }
    return new IdentityCredentialsProvider(authentication);
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java [hunk 4]
```diff
@@ -175,43 +277,86 @@ public ScheduledExecutorService jobWorkerExecutor() {
...
+    // do not use any port
+    return camundaClientProperties.getZeebe().getGatewayUrl().getHost();
+  }
 
-        @Override
-        public boolean shouldRetryRequest(final StatusCode statusCode) {
-          return statusCode.code() == Code.DEADLINE_EXCEEDED.value();
-        }
-      };
+  private boolean composePlaintext() {
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:341]: 🔧 

```suggestion
  private boolean composePlaintext() {
    String protocol = camundaClientProperties.getZeebe().getGatewayUrl().getProtocol();
    switch (protocol) {
        case "http":
            return true;
        case "https":
            return false;
        default:
            throw new IllegalStateException(
                    String.format("Unrecognized zeebe protocol '%s'", protocol));
    }
}
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java [hunk 5]
```diff
@@ -175,43 +277,86 @@ public ScheduledExecutorService jobWorkerExecutor() {
...
 
-    if (commonConfigurationProperties.getEnabled()
-        && !(authentication instanceof DefaultNoopAuthentication)) {
-      return new CredentialsProvider() {
-        @Override
-        public void applyCredentials(final CredentialsApplier applier) {
-          final Map.Entry<String, String> authHeader = authentication.getTokenHeader(Product.ZEEBE);
-          applier.put(authHeader.getKey(), authHeader.getValue());
-        }
+  private String composeGatewayAddress() {
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:315]: 🔧 

```suggestion
  private String composeGatewayAddress() {
    ZeebeGatewayUrl gatewayUrl = camundaClientProperties.getZeebe().getGatewayUrl();
    int port = gatewayUrl.getPort();
    int defaultPort = gatewayUrl.getDefaultPort();
    String host = gatewayUrl.getHost();

    if (port != -1) {
        return composeAddressWithPort(host, port, "Gateway port is set");
    } else if (defaultPort != -1) {
        return composeAddressWithPort(host, defaultPort, "Gateway port has default");
    } else {
        LOG.debug("Gateway cannot be determined, address will be '{}'", host);
        return host;
    }
}

private String composeAddressWithPort(String host, int port, String debugMessage) {
    String gatewayAddress = host + ":" + port;
    LOG.debug(debugMessage + ", address will be '{}'", gatewayAddress);
    return gatewayAddress;
}
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java [hunk 6]
```diff
@@ -175,43 +277,86 @@ public ScheduledExecutorService jobWorkerExecutor() {
...
+    throw new IllegalStateException(
+        String.format(
+            "Unrecognized zeebe protocol '%s'",
+            camundaClientProperties.getZeebe().getGatewayUrl().getProtocol()));
+  }
+
+  private CredentialsProvider legacyCredentialsProvider() {
     if (hasText(properties.getCloud().getClientId())
         && hasText(properties.getCloud().getClientSecret())) {
+      //        log.debug("Client ID and secret are configured. Creating OAuthCredientialsProvider
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:358]: 🔧 Can we remove those commented lines ?

File: spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      "camunda.client.mode=saas",
+      "camunda.client.cluster-id=12345",
+      "camunda.client.region=bru-2",
+      "camunda.client.auth.client-id=my-client-id",
+      "camunda.client.auth.client-secret=my-client-secret",
+      "camunda.client.auth.issuer=http://localhost:14682/auth-server"
+    })
+@WireMockTest(httpPort = 14682)
+public class AuthenticationConfigurationSaasTest {
+  private static final String ACCESS_TOKEN =
```
Comment (nicpuppa) [inline spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java:50]: ❌ Please remove this token, you can use a random string

**Review by nicpuppa — COMMENTED** [review]
Hey @ana-vinogradova-camunda great work so far. I made a first round of review. Please have a look 👀

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/CommonClientConfiguration.java (hunk 1 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/CommonClientConfiguration.java:49]: So customers can provide their own definition for bean and in such case this method will be completely ignored.

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java (hunk 2 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java:38]: I wouldn't remove` try - catch` blocks as we don't want to throw exceptions here, we just want to move to the next code block.

Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java:38]: ` if (property == null)` then it we want to check if we can get legacy properties, if not - try to get properties from the supplier and if that fails - return default properties.
I might be missing something, suggestions are welcome 👍

Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java:38]: I see, thanks for the suggestion 👍

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java (hunk 4 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:341]: Introduced `switch` in my latest commit, thank you 👍

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java (hunk 5 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfigurationSpringImpl.java:316]: Applied these changes in my latest commit, thanks 👍

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java (hunk 2 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/PropertyUtil.java:38]: Applied the changes, thanks!

**Review by nicpuppa — APPROVED** [review]
LGTM 🚀

**Pull request merged**